File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/application/exception/ProductUnavailableException.java
package com.simple.ecommerce.application.exception;

import com.simple.ecommerce.domain.entity.Product;

/**
 * Exception thrown when trying to add an unavailable product to the shopping cart.
 */
public class ProductUnavailableException extends RuntimeException {

    private final Product product;

    public ProductUnavailableException(Product product) {
        super("Product '" + product.getName() + "' is not available for purchase");
        this.product = product;
    }

    public ProductUnavailableException(String message, Product product) {
        super(message);
        this.product = product;
    }

    /**
     * Gets the unavailable product.
     *
     * @return the product that is unavailable
     */
    public Product getProduct() {
        return product;
    }
}

File: Projects/Testtrack/p911 - Simple eCommerce backend/application/exception/product_unavailable_exception.py
from typing import Optional
from domain import Product


class ProductUnavailableException(Exception):
    def __init__(self, product: Product, message: Optional[str] = None):
        if message is None:
            message = f"Product '{product.get_name()}' is not available for purchase"
        super().__init__(message)
        self._product = product

    def get_product(self) -> Product:
        return self._product
